### PR TITLE
feat: allow slicing by tuple of sequences of ints

### PR DESCRIPTION
**Pull request by MarcoGorelli**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
@EdAbati fancy taking a look, as you did quite some work on slicing?

File: narwhals/_pandas_like/dataframe.py [hunk 1]
```diff
@@ -122,11 +122,30 @@ def __getitem__(
...
+            )
+            raise TypeError(msg)  # pragma: no cover
+
         elif isinstance(item, tuple) and len(item) == 2:
             from narwhals._pandas_like.series import PandasLikeSeries
 
             if isinstance(item[1], str):
-                native_series = self._native_dataframe.loc[item]
+                item = (item[0], self._native_dataframe.columns.get_loc(item[1]))
+                native_series = self._native_dataframe.iloc[item]
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/dataframe.py:148]: drive-by: this wasn't quite right to begin with. I've added an extra test case to `test_gather_pandas_index` which demonstrates the issue (which this pr fixes)

no blame of course 🤗 this is just a fyi

File: narwhals/dataframe.py [hunk 2]
```diff
@@ -449,7 +449,13 @@ def get_column(self, name: str) -> Series:
...
 
     @overload
-    def __getitem__(self, item: tuple[Sequence[int], str | int]) -> Series: ...  # type: ignore[overload-overlap]
+    def __getitem__(self, item: tuple[Sequence[int], Sequence[int]]) -> Self: ...
+    @overload
+    def __getitem__(self, item: tuple[Sequence[int], str]) -> Series: ...  # type: ignore[overload-overlap]
+    @overload
+    def __getitem__(self, item: tuple[Sequence[int], Sequence[str]]) -> Self: ...
+    @overload
+    def __getitem__(self, item: tuple[Sequence[int], int]) -> Series: ...  # type: ignore[overload-overlap]
```
Comment (MarcoGorelli) [inline narwhals/dataframe.py:458]: this required some overload ignores, but overall, it looks correct:
```
dfpl = pl.DataFrame({'a': [1,2,3], 'b': [4,5,6]})

df = nw.from_native(dfpl, eager_only=True)

reveal_type(df[[0, 1], 'a'])
reveal_type(df[[0, 1], 0])
reveal_type(df[[0, 1], [0, 1]])
reveal_type(df[[0, 1], ['a', 'b']])
```
mypy reveals the first two to be Series, and the last two to be DataFrame

**Review by EdAbati — APPROVED** [review]
Looks good to me👌👌

Regarding typing, a part from the question below, there are some overload missing in `_arrow/dataframe` and `_pandas_like/dataframe`. Since mypy is not picking them up maybe they are not 100% required though, what do you think?

File: narwhals/stable/v1.py
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply narwhals/stable/v1.py:1]: it's only the overloads that are user-facing anyway, so I wouldn't worry too much about it

**Comment by MarcoGorelli** [comment]
Thanks for your review 🙏 !

`_pandas_like` and `_arrow` aren't user-facing - it's better if we get the types there for internal correctness, but I'm not stressing out over them too much 😄
